### PR TITLE
Ability to delete a team

### PR DESCRIPTION
**Pull request by slorek**
https://trello.com/c/FRRtOhuq/668-5-absorb-opss-product-safety-enforcement-into-opss-enforcement

## Description
This allows us to delete a team from service.

When a team is deleted:

- all users on that team are transferred to another designated team. Any users that own a case retain ownership. All user roles are transferred to the new team
- all case collaborations for that team are transferred to another designated team. If the team owns any cases, the designated team will inherit ownership.
- audit activity is generated as normal - except when a user on the deleted team is the owner of a case. In this event we add an extra activity showing the team ownership has changed
- no emails are sent
- the team will no longer be in the filter cases list
- the team will no longer appear in the list to transfer case ownership
- the team will no longer appear in the list to add a team to the case

In our initial use case this process would ordinarily generate many hundreds of emails to the affected users. This implementation suppresses those emails. This could be made optional in a future iteration.

## Checklist:
- [x] Have you documented your changes in the pull request description?
- [ ] Does the change present any security considerations?
- [ ] Is any gem functionality overloaded? Eg: Devise controller methods being overloaded.
- [x] Has acceptance criteria been tested by a peer?
- [ ] Has the CHANGELOG been updated? (If change is worth telling users about.)

### General testing (author)
- [ ] Test without JavaScript
- [ ] Test on small screen

### Accessibility testing (author)
- [ ] Works keyboard only
- [ ] Tested with one screen reader
- [ ] Zoom page to 400% - content still visible
- [ ] Disable CSS - does content make sense and appear in a logical order?

### CODE REVIEW

File: .rubocop.yml [hunk 1]
```diff
@@ -36,6 +36,9 @@ Style/TrailingCommaInArguments:
 Rails/HelperInstanceVariable:
   Enabled: false
 
+Lint/AmbiguousBlockAssociation:
```
Comment (slorek) [inline .rubocop.yml:39]: Disabled this because fixing it introduces syntax errors: https://github.com/rubocop-hq/rubocop/issues/4222

File: app/helpers/investigations/user_filters_helper.rb [hunk 2]
```diff
@@ -1,6 +1,6 @@
 module Investigations::UserFiltersHelper
   def entities
-    User.get_owners(except: current_user).decorate + Team.all_with_organisation.decorate
```
Comment (slorek) [inline app/helpers/investigations/user_filters_helper.rb:3]: We no longer need the team organisation. This was previously used in the `display_name` method but was changed a while ago to no longer require it.

File: app/services/delete_team.rb [hunk 3]
```diff
@@ -0,0 +1,109 @@
+class DeleteTeam
+  include Interactor
+
+  delegate :team, :new_team, :user, to: :context
```
Comment (frankieroberto) [inline app/services/delete_team.rb:4]: Suggest rewording this to be clearer that it’s deleting a team and merging all its users & cases into another team:

```suggestion
class MergeTeams
  include Interactor

  delegate :team_to_delete, :team_to_keep, :user, to: :context
```

File: app/services/delete_team.rb [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    metadata = update_owner_activity_class.build_metadata(new_team, change_case_owner_rationale)
+
+    update_owner_activity_class.create!(
+      source: UserSource.new(user: user),
+      investigation: collaboration.investigation,
+      title: nil,
+      body: nil,
+      metadata: metadata
+    )
+  end
```
Comment (frankieroberto) [inline app/services/delete_team.rb:53]: These method names are so similar it's kinda hard to follow. I _think_ I understand what’s going on, but it's taken a bit of head-scratching. I don't have any immediate suggestions on how to make it clearer though. 😬

File: db/migrate/20200911142816_add_deleted_at_to_teams.rb [hunk 5]
```diff
@@ -0,0 +1,8 @@
+class AddDeletedAtToTeams < ActiveRecord::Migration[5.2]
+  disable_ddl_transaction!
+
+  def change
+    add_column :teams, :deleted_at, :datetime
+    add_index :teams, :deleted_at, algorithm: :concurrently
```
Comment (frankieroberto) [inline db/migrate/20200911142816_add_deleted_at_to_teams.rb:6]: I wonder if a partial index would be helpful here, as we'd normally be querying for non-deleted teams?

```suggestion
    add_index :teams, :deleted_at, name: 'not_deleted_index', where: {deleted_at: nil), algorithm: :concurrently
```

(Probably marginal gains at our scale though)

File: spec/features/add_team_to_case_spec.rb [hunk 6]
```diff
@@ -46,6 +47,10 @@
     expect(page).to have_selector("a", text: "Select the permission level the team should have")
     expect(page).to have_selector("a", text: "Select whether you want to include a message")
 
+    # Check deleted teams are not listed
+    expect(page).to have_css("#team option[value=\"#{team.id}\"]")
+    expect(page).not_to have_css("#team option[value=\"#{deleted_team.id}\"]")
```
Comment (frankieroberto) [inline spec/features/add_team_to_case_spec.rb:52]: You could use the Capybara method here to check the values of the select?

```suggestion
    # Check deleted teams are not listed
    expect(page).not_to have_select("Select other team name", with_options: [deleted_team.name])
    expect(page).to have_select("Select other team name", with_options: [team.name])
```

**Review by frankieroberto — APPROVED** [review]

File: app/services/delete_team.rb (hunk 4 above)
Comment (slorek) [reply app/services/delete_team.rb:53]: Agree.

**Review by nasirkhanbeis — APPROVED** [review]
looks good as expected!

File: app/services/delete_team.rb (hunk 3 above)
Comment (slorek) [reply app/services/delete_team.rb:4]: 🤔 not sure about this. The merging is the thing you have to do to accomplish the team deletion.

File: app/services/delete_team.rb [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+  include Interactor
+
+  delegate :team, :new_team, :user, to: :context
+
+  def call
+    context.fail!(error: "No team supplied") unless team.is_a?(Team)
+    context.fail!(error: "No new team supplied to absorb cases and users") unless new_team.is_a?(Team)
+    context.fail!(error: "No user supplied") unless user.is_a?(User)
+    context.fail!(error: "Team cannot already be deleted") if team.deleted?
+    context.fail!(error: "New team cannot already be deleted") if new_team.deleted?
```
Comment (StupidCodeFactory) [inline app/services/delete_team.rb:11]:
```suggestion
    context.fail!(error: "New team already be deleted") if new_team.deleted?
```

File: app/services/delete_team.rb [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+  def remove_team_as_collaborator_from_cases
+    team.collaboration_accesses.changeable.each do |collaboration|
+      add_new_team_to_case(collaboration) unless new_team_already_collaborating?(collaboration.investigation)
+      remove_team_from_case(collaboration)
+    end
+  end
+
+  def new_team_already_collaborating?(investigation)
+    investigation.collaboration_accesses.find_by(collaborator_id: new_team.id).present?
```
Comment (StupidCodeFactory) [inline app/services/delete_team.rb:67]:
```suggestion
    investigation.collaboration_accesses.where(collaborator: new_team).exists?
```

File: lib/tasks/delete_team.rake [hunk 9]
```diff
@@ -0,0 +1,15 @@
+namespace :team do
+  desc "Marks the given team as deleted, assigning their cases and users to another team"
+  task delete: :environment do
+    team = Team.find(ENV.fetch("ID", nil))
```
Comment (StupidCodeFactory) [inline lib/tasks/delete_team.rake:4]:
```suggestion
    team = Team.find(ENV["ID"])
```

File: doc/tasks.md [hunk 10]
```diff
@@ -23,3 +23,11 @@ cf run-task psd-web "export \$(./env/get-env-from-vcap.sh) && ID=<user ID> rake
...
 
 Where `EMAIL` or `ID` is the e-mail address or ID of the user to be deleted. Task name can be set to anything you like.
+
+## Deleting a Team
+
+```bash
+cf run-task psd-web "export \$(./env/get-env-from-vcap.sh) && ID=<id> NEW_TEAM_ID=<id> EMAIL=<email address> rake team:delete" --name <task name>
+```
+
+Case collaborations and users which belong to the team identified by `ID` will be migrated to another team identified by `NEW_TEAM`. The user identified by `EMAIL` will be attributed to the change on all relevant audit activity.
```
Comment (StupidCodeFactory) [inline doc/tasks.md:33]:
```suggestion
Case collaborations and users which belong to the team identified by `ID` will be migrated to another team identified by `NEW_TEAM_ID`. The user identified by `EMAIL` will be attributed to the change on all relevant audit activity.
```

File: lib/tasks/delete_team.rake [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+    new_team = Team.find(ENV.fetch("NEW_TEAM_ID", nil))
+    user = User.find_by!(email: ENV.fetch("EMAIL", nil))
```
Comment (StupidCodeFactory) [inline lib/tasks/delete_team.rake:6]:
```suggestion
    user = User.find_by!(email: ENV["EMAIL"]))
```

File: lib/tasks/delete_team.rake [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+
+    result = DeleteTeam.call(team: team, new_team: new_team, user: user)
```
Comment (StupidCodeFactory) [inline lib/tasks/delete_team.rake:8]:
```suggestion
    result = DeleteTeam.call!(team: team, new_team: new_team, user: user)
```
and remove the line below. It should be the same.

File: spec/features/filter_investigations_spec.rb [hunk 13]
```diff
@@ -33,12 +34,18 @@
...
+    expect(page).to have_css("#case_owner_is_someone_else_id option[value=\"#{other_team.id}\"]")
     expect(page).to have_css("#case_owner_is_someone_else_id option[value=\"#{another_active_user.id}\"]")
     expect(page).not_to have_css("#case_owner_is_someone_else_id option[value=\"#{another_inactive_user.id}\"]")
+    expect(page).not_to have_css("#case_owner_is_someone_else_id option[value=\"#{other_deleted_team.id}\"]")
 
+    expect(page).to have_css("#created_by_someone_else_id option[value=\"#{team.id}\"]")
+    expect(page).to have_css("#created_by_someone_else_id option[value=\"#{other_team.id}\"]")
     expect(page).to have_css("#created_by_someone_else_id option[value=\"#{another_active_user.id}\"]")
     expect(page).not_to have_css("#created_by_someone_else_id option[value=\"#{another_inactive_user.id}\"]")
+    expect(page).not_to have_css("#created_by_someone_else_id option[value=\"#{other_deleted_team.id}\"]")
```
Comment (StupidCodeFactory) [inline spec/features/filter_investigations_spec.rb:48]: Same as @frankieroberto comment above, you could use the `have_select(team.name)`, that would also test that the value in the drop down is the one team name.

File: spec/services/delete_team_spec.rb [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+      end
+
+      it "marks the team as deleted" do
+        freeze_time do
+          expect { delete_team }.to change(team, :deleted_at).from(nil).to(Time.zone.now)
+        end
+      end
+
+      it "migrates the users to the new team" do
+        expect { delete_team }.to change { team_user.reload.team }.from(team).to(new_team)
```
Comment (StupidCodeFactory) [inline spec/services/delete_team_spec.rb:79]:
```suggestion
        expect { delete_team }.to change(team_user, :team).from(team).to(new_team)
```

File: spec/services/delete_team_spec.rb [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+        before do
+          ChangeCaseOwner.call!(
+            investigation: team_case,
+            owner: team,
+            user: team_user
+          )
+        end
+
+        it "transfers ownership of the case to the new team" do
+          expect { delete_team }.to change { team_case.reload.owner }.from(team).to(new_team)
```
Comment (StupidCodeFactory) [inline spec/services/delete_team_spec.rb:113]:
```suggestion
          expect { delete_team }.to change(team_case, :owner).from(team).to(new_team)
```

File: spec/services/delete_team_spec.rb [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+          expect { delete_team }.not_to have_enqueued_mail(NotifyMailer, :investigation_updated)
+          expect { delete_team }.not_to have_enqueued_mail(NotifyMailer, :team_deleted_from_case_email)
+        end
+      end
+
+      context "when the team is a collaborator on a case owned by the new team" do
+        let!(:new_team_case) { create(:allegation, creator: new_team_user, read_only_teams: [team]) }
+
+        it "removes the team from the case" do
+          expect { delete_team }.to change { new_team_case.reload.teams_with_read_only_access }.from([team]).to([])
```
Comment (StupidCodeFactory) [inline spec/services/delete_team_spec.rb:170]:
```suggestion
          expect { delete_team }.to change(new_team_case, :teams_with_read_only_access).from([team]).to([])
```

File: spec/services/delete_team_spec.rb [hunk 17]
```diff
@@ -0,0 +1,254 @@
...
+
+        it "does not send notification e-mails", :with_test_queue_adapter, :aggregate_failures do
+          expect { delete_team }.not_to have_enqueued_mail(NotifyMailer, :team_added_to_case_email)
+          expect { delete_team }.not_to have_enqueued_mail(NotifyMailer, :team_deleted_from_case_email)
+          expect { delete_team }.not_to have_enqueued_mail(NotifyMailer, :case_permission_changed_for_team)
+        end
+
+        context "when the new team is not already a collaborator on the case" do
+          it "adds the new team to the case with the same access level as the old team" do
+            expect { delete_team }.to change { other_team_case.reload.teams_with_read_only_access }.from([team]).to([new_team])
```
Comment (StupidCodeFactory) [inline spec/services/delete_team_spec.rb:209]:
```suggestion
            expect { delete_team }.to change(other_team_case, :teams_with_read_only_access).from([team]).to([new_team])
```

**Review by StupidCodeFactory — APPROVED** [review]

File: spec/services/delete_team_spec.rb (hunk 17 above)
Comment (slorek) [reply spec/services/delete_team_spec.rb:209]: Doesn't work. Need `reload`.

File: spec/services/delete_team_spec.rb (hunk 16 above)
Comment (slorek) [reply spec/services/delete_team_spec.rb:170]: Doesn't work. Need `reload`.

File: spec/services/delete_team_spec.rb (hunk 15 above)
Comment (slorek) [reply spec/services/delete_team_spec.rb:113]: Doesn't work. Need `reload`.

File: spec/services/delete_team_spec.rb (hunk 14 above)
Comment (slorek) [reply spec/services/delete_team_spec.rb:79]: Doesn't work. Need `reload`.

File: spec/features/filter_investigations_spec.rb (hunk 13 above)
Comment (slorek) [reply spec/features/filter_investigations_spec.rb:48]: I refactored the other existing expectations as well, looks a lot better.

File: lib/tasks/delete_team.rake (hunk 12 above)
Comment (slorek) [reply lib/tasks/delete_team.rake:8]: It's not quite the same:

```ruby
rake aborted!
Interactor::Failure: #<Interactor::Context team=#<Team id: "01a0e8df-5d08-4a78-a796-883c82f00aa8", created_at: "2020-06-04 15:53:26", name: "OPSS Trading Standards Co-ordination", organisation_id: "771d91ab-0da2-46e7-93c4-b6567842c2af", team_recipient_email: nil, updated_at: "2020-09-21 11:00:16", deleted_at: "2020-09-21 11:00:16">, new_team=#<Team id: "9c127846-1ccc-4182-95c6-6c9044c23feb", created_at: "2020-06-04 15:53:26", name: "OPSS Incident Management", organisation_id: "771d91ab-0da2-46e7-93c4-b6567842c2af", team_recipient_email: nil, updated_at: "2020-06-04 15:53:26", deleted_at: nil>, user=#<User id: "618126d3-e7af-4765-9e4b-a732e438b534", account_activated: true, created_at: "2020-06-04 15:53:27", credential_type: nil, direct_otp: "98208", direct_otp_sent_at: "2020-08-20 13:47:39", email: "admin@example.com", encrypted_otp_secret_key: nil, encrypted_otp_secret_key_iv: nil, encrypted_otp_secret_key_salt: nil, has_accepted_declaration: true, has_been_sent_welcome_email: true, has_viewed_introduction: false, hash_iterations: 27500, invitation_token: "829f4b7200d1d5dc26012e5a990f77", invited_at: "2020-06-04 15:53:27", keycloak_created_at: nil, mobile_number: "07791702127", mobile_number_verified: true, name: "Team Admin", organisation_id: "771d91ab-0da2-46e7-93c4-b6567842c2af", second_factor_attempts_count: 1, second_factor_attempts_locked_at: nil, updated_at: "2020-09-22 13:54:11", secondary_authentication_operation: "secondary_authentication", deleted_at: nil, team_id: "10e6cee6-7a7d-431e-aefb-8dfe9c1123fe", skip_password_validation: false>, error="Team cannot already be deleted">
```
versus:

```
rake aborted!
Team cannot already be deleted
```
